data/tutorials/language/0lg_10_polymorphic_variants.md

```diff
@@ -0,0 +1,713 @@
+
+**Prerequisites**: This is an intermediate-level tutorial. It is required to have completed tutorials on [Functions and Values](/docs/functions-and-values), [Basic Data Types](/docs/basic-data-types), and [Lists](/docs/lists) to begin this one.
+
+### Origin and Context
+
+Polymorphic variants originate from Jacques Garrigue's work on Objective Label, which was [first published in 1996](https://caml.inria.fr/pub/old_caml_site/caml-list-ar/0533.html). It became part of standard Objective Caml with [release 3.0](https://caml.inria.fr/distrib/ocaml-3.00/) in 2000, along with labelled and optional function arguments. They were introduced to give more precise types in [LablTk](https://garrigue.github.io/labltk/).
+
+The core type system of OCaml follows a [_nominal_](https://en.wikipedia.org/wiki/Nominal_type_system) discipline. Variants must be explicitly declared before being used. The typing discipline used for polymorphic variants and classes is different, as it is [_structural_](https://en.wikipedia.org/wiki/Structural_type_system).
+
+In the nominal approach of typing, types are first defined; later, when the type of an expression is inferred, three outcomes are possible:
+1. If a matching type is found, it becomes the inferred type.
+1. If any type can be applied, a type parameter is created.
+1. If typing inconsistencies are found, an error is raised.
+
+This is very similar to solving an equation in mathematics. Equations accept either zero, exactly one, several, or infinitely many numbers as solutions. Nominal type-checking finds that either zero, exactly one, or any type can be used in an expression.
+
+In the structural approach of typing, type definitions are optional, so they can be omitted. Type-checking an expression constructs a data structure that represents the types that are compatible with it. These data structures are displayed as type expressions sharing a resemblance with simple variants.
+
+<!--
+### Learning Goals
+
+The goal of this tutorial is to make the reader acquire the following capabilities:
+- Write polymorphic variants using code, from scratch, using mainstream features
+- Maintain existing code that is using polymorphic variants
+- Sort out polymorphic variant types and errors
+  - Add type annotation needed to resolve polymorphic variant type-checking issues
+  - Add or remove catch-all patterns in polymorphic variant pattern matching
+  - Insert or remove type casts
+- Refactor between simple and polymorphic variants
+- Choose to use polymorphic variants when really needed
+-->
+
+### A Note on Simple Variants and Polymorphism
```

Comment:
Looking at the learning goals for this page, I'm not sure if this clarification helps me write/use/assess polyvars at all.

Feels like a "nice to know" but not a "must know".

Reply:
I agree, if this section belongs to this tutorial, it is not at its place.

Reply:
Your suggestion makes sense too. I'll do that.